### DISCUSSION TITLE
Implementing captcha for public submissions

### DISCUSSION THREAD

**Discussion by pwalsh** [govflow/govflow#32] [Ideas]
We want to reduce/eliminate spam at public service request forms.

The general approach is to use a captcha service: primarily Google's [reCAPTCHA](https://developers.google.com/recaptcha), or alternatively [hCaptcha](https://www.hcaptcha.com) (Cloudflare use hcaptcha FTR).

# Implementation

Both captcha services above have more or less the same implementation, being:

  - integrate some javascript on the client to inject a captcha widget into one or many forms
  - on submit of the form, check a special field server side, and with the value in that field, hit an api endpoint to verify if the submitted value is valid
     - if valid, let rest of form submit as normal
     - if not valid, send back to the client and get the user to complete the capture

As we want to support Open311's GeoReport V2, we have an endpoint to POST service requests in that format, and requiring any type of captcha service "breaks" the contract with Open311 clients submitting to the server (as Open311 is designed so that theoretically any compliant form can submit to any compliant server).

In practice though, it is reasonable to assume that anyone who runs Gov Flow with Open311 integration does in fact want to eliminate abuse of the system, so I think the following solution is workable:

- Any Open311 endpoint that allows writing data is by default required to negotiate a captcha (for web clients), or, an authz token (for programmatic integration). Obviously the request handler will do the auth/captcha negotiation and accept valid Open311 payloads only after that. There can be a fallback mechanism between captcha / authz (authz'd clients can bypass captcha).
- A Gov Flow administrator can **opt out** of captcha and/or authz for Open311 endpoints.

For clarity, the write endpoints we have that are **not** for Open311 compat. should be treated a bit differently:

- They require captcha and/or auth
- We shouldn't offer an opt-out (it can be done with plugins/code, but, then it is in the hands of the implementor)
- Maybe related, we should focus our client implementations on forms for these endpoints, reserving Open311-compatible forms for the most basic use case, where Open311 combat. is required/desired.

# Estimates

The server integration, including opt-out functionality, is probably around a day of work. I guess around the same for the basic client integration.

**But note**: there is kind of an implied feature here, in that the configuration of the server in certain scenarios needs to be exposed for our client forms to consume, so when they init themselves, they can know, for example, if captcha is required or not. This could be (i) a simple endpoint with some boolean flags which the client has hard-coded handling of, or, (ii) we could start to move to a declarative forms approach (I wrote some [notes about it here](https://github.com/govflow/govflow/issues/6)) where the client initializes its fields and behavior based on a schema the server provides. Option (ii) is much more complex in the short run, and deserves a spec of its own, but I imagine the more customization we allow (relationship between external workflows and public request forms, for example) the stronger the need for this type of approach will be.

# References

Some background at #7 and #8 and #6

**Comment by idoivri** [comment]
to your point on declarative forms, I think it's generally where we want to go: extra flexibility for our clients to define what data they need from residents.

**Comment by amirozer** [comment]
Looks good. I agree with the approach of making captcha optional and overridable when authenticated.

Regarding declarative forms, I'm reminding that Open311 has its own way of declaring the form schema, and our form already supports it. If at some point we decide we're too limited by Open311 we can create an alternative endpoint.

I suggest to start without having a backend endpoint to return the config, and instead requiring turning on Captcha when integrating the form.

**Comment by pwalsh** [comment]
@amirozer 

After further review at https://developers.google.com/recaptcha/docs/versions and also the promo video at https://developers.google.com/recaptcha I think we should use v2 and not v3. Reasoning is that v3 generally requires deeper interaction of the user on your site in order to create accurate scores, but in our use case, users will hit a standalone form with no history of interaction, so, it seems it is best to get the user to validate herself with a challenge.

**Reply by noamoss** [reply]
Assuming some of the users (workers) will return again and again to the form (unless they prefer submitting an email) - does it change anything about the captcha version?

**Comment by pwalsh** [comment]
This is now implemented.
